File: javascript/056.js
//  056
//  Merge Intervals
//  2015-12-13
/////////////////////////////////////////////////////
/**
 * Definition for an interval.
 */
function Interval(start, end) {
    this.start = start;
    this.end = end;
}
/**
 * @param {Interval[]} intervals
 * @return {Interval[]}
 */
var merge = function(intervals) {
    intervals.sort(function(a,b) {return a.start-b.start;});
    if (intervals.length == 0)
        return [];
    var r = [intervals[0]];
    for (var i = 1; i < intervals.length; i++)
        if (intervals[i].start <= r[r.length-1].end)
            r[r.length-1].end = Math.max(intervals[i].end, r[r.length-1].end);
        else
            r.push(intervals[i]);
    return r;
};

/////////////////////////////////////////////////////
var intervals = [
    new Interval(2,6),
    new Interval(1,3),
    new Interval(8,10),
    new Interval(15,18)
/*
    new Interval(1,4),
    new Interval(2,3)

    new Interval(2,3),
    new Interval(4,5),
    new Interval(6,7),
    new Interval(8,9),
    new Interval(1,10)

    new Interval(1,4),
    new Interval(0,2),
    new Interval(3,5)
*/
];

var r = merge(intervals);
console.log(r);

File: cpp/001_100/056.cpp
//
//  056.cpp
//  leetcode
//
//  Created by Hang Zhang on 12/9/15.
//  Copyright (c) 2015 Hilings Studio. All rights reserved.
//

#include <iostream>
#include <vector>
using namespace std;

/**
 * Definition for an interval.
 */
struct Interval {
    int start;
    int end;
    Interval() : start(0), end(0) {}
    Interval(int s, int e) : start(s), end(e) {}
};

void show(Interval a) {
    cout << "[" << a.start << ", " << a.end << "]\n";
}

bool cmp (Interval a, Interval b) {
    return a.start < b.start;
}

class Solution {
public:
    vector<Interval> merge(vector<Interval>& intervals) {
        vector<Interval> v;
        if (intervals.empty())
            return v;
        sort(intervals.begin(), intervals.end(), cmp);

        while (intervals.size() >= 2) {
            int index;
            for (index = (int)intervals.size()-2; index >= 0; index--) {
                if (intervals.back().start <= intervals[index].end)
                    break;
            }
            if (index < 0) {
                v.push_back(intervals.back());
                intervals.pop_back();
            } else {
                int lower = intervals[index].start;
                int upper = max(intervals[index].end, intervals.back().end);
                for (int i = (int)intervals.size()-1; i >= index; i--) {
                    intervals.pop_back();
                }
                intervals.push_back(Interval(lower, upper));
            }
        }

        v.push_back(intervals.front());
        return v;
    }
};


int main(int arg, char *argv[]) {
    // insert code here...
    cout << "LeetCode 056 Merge Intervals, C++ ...\n\n";

    vector<Interval> intervals {
        ///*
        Interval(2,6),
        Interval(1,3),
        Interval(8,10),
        Interval(15,18)
         //*/

        /*
        Interval(1,4),
        Interval(2,3)
         */

        /*
        Interval(2,3),
        Interval(4,5),
        Interval(6,7),
        Interval(8,9),
        Interval(1,10)
         */
    };

    //return [1,6],[8,10],[15,18]
    for (int i = 0; i < intervals.size(); i++) {
        //show(intervals[i]);
    }

    Solution sol;
    vector<Interval> r = sol.merge(intervals);
    for (int i = 0; i < r.size(); i++) {
        show(r[i]);
    }

    return 0;
}